Migrate LST inputs to SoA collections

ariostas opened the pull request:
This PR refactors the LST inputs so that Portable SoA Collections are used from the very beginning. It also deduplicates some code since there were separate functions for standalone and CMSSW that played the same role.

It's mostly done, but we have a couple of questions (probably for @fwyzard or @makortel).

1. Running on CPU works fine, but I'm still having issues with GPUs. The issue is with LCG dictionaries, as I described in [my Mattermost message](https://mattermost.web.cern.ch/cms-o-and-c/pl/jo9f68nzwbnmmkh38dr3qtf3fa). Initially, I thought that the ROOT name normalization issue that Matti mentioned could be the cause, but I think it's something else.
2. I had to split an SoA collection into two, since part of it is created before LST and some needs to be modified during the algorithm (to add some extra data). Do you know of a good way to combine the two collections into some sort of combined view, or do we just need to keep using the two separately. (Obviously, we could just define some new struct or class, but that doesn't sound great to me)

c.c. @slava77 @VourMa

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824":{"files":["RecoTracker/LST/plugins/alpaka/LSTProducer.cc","RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h","RecoTracker/LSTCore/interface/LSTPrepareInput.h","RecoTracker/LSTCore/interface/alpaka/LST.h","RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h","RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.h","RecoTracker/LSTCore/src/alpaka/classes_cuda.h","RecoTracker/LSTCore/src/alpaka/classes_cuda_def.xml","RecoTracker/LSTCore/src/alpaka/classes_rocm.h","RecoTracker/LSTCore/src/alpaka/classes_rocm_def.xml"],"squashed":false,"time":1743085003},"0bb4f24862a5cd6f76b923387166e5931d6e5bdb":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LST/plugins/alpaka/LSTProducer.cc","RecoTracker/LSTCore/interface/LSTInput.h"],"squashed":false,"time":1742220462},"38eca3068bf7842ca155d5e6894a191af0074aef":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LSTCore/interface/LSTInput.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.h","RecoTracker/LSTCore/src/alpaka/Segment.h","RecoTracker/LSTCore/standalone/bin/lst.cc"],"squashed":false,"time":1742214469},"5cc987239cae28df23baaeb1855de57a2ac5fb27":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LSTCore/interface/Common.h","RecoTracker/LSTCore/interface/LSTInputSoA.h","RecoTracker/LSTCore/interface/LSTPrepareInput.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/Segment.h","RecoTracker/LSTCore/standalone/bin/lst.cc"],"squashed":false,"time":1742829246},"69a896d6bafbe4aeba5934126a6b3788be1a09cf":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LSTCore/interface/LSTPrepareInput.h","RecoTracker/LSTCore/src/alpaka/Hit.h","RecoTracker/LSTCore/standalone/bin/lst.cc"],"squashed":false,"time":1742913402},"799c1937743184ee6a0e8d2a5b37997eafd90a61":{"files":["RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h","RecoTracker/LSTCore/interface/LSTInputSoA.h","RecoTracker/LSTCore/interface/LSTPrepareInput.h","RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.h","RecoTracker/LSTCore/src/classes.cc","RecoTracker/LSTCore/standalone/Makefile","RecoTracker/LSTCore/standalone/bin/lst.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.h"],"squashed":false,"time":1743707354},"7e63a0995e3ff59727ac683c7bcbbffda69bf0e7":{"files":["RecoTracker/LST/src/classes_def.xml","RecoTracker/LSTCore/interface/LSTInputSoA.h"],"squashed":false,"time":1742388581},"9330ea30dd75eadf4055a26ba3310d4d6614c861":{"files":["RecoTracker/LST/src/classes.h","RecoTracker/LST/src/classes_def.xml","RecoTracker/LSTCore/BuildFile.xml","RecoTracker/LSTCore/interface/LSTInputHostCollection.h","RecoTracker/LSTCore/src/classes.cc","RecoTracker/LSTCore/src/classes.h","RecoTracker/LSTCore/src/classes_def.xml"],"squashed":false,"time":1742994392},"95de7e001bbb8bf56603a8ebab9329868cfa8a98":{"files":["RecoTracker/IterativeTracking/python/HighPtTripletStep_cff.py","RecoTracker/LST/plugins/alpaka/LSTProducer.cc","RecoTracker/LST/python/lst_cff.py"],"squashed":false,"time":1742219247},"a3b6a7fb36e84bb1ad522934380e04657b311866":{"files":["RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc"],"squashed":false,"time":1742214406},"b39f14dccdb2e79ee802a2c46e0325db31ef9150":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LST/plugins/alpaka/LSTProducer.cc","RecoTracker/LSTCore/interface/HitsHostCollection.h","RecoTracker/LSTCore/interface/HitsRangesHostCollection.h","RecoTracker/LSTCore/interface/HitsSoA.h","RecoTracker/LSTCore/interface/LSTInput.h","RecoTracker/LSTCore/interface/LSTInputHostCollection.h","RecoTracker/LSTCore/interface/LSTInputSoA.h","RecoTracker/LSTCore/interface/PixelSegmentsSoA.h","RecoTracker/LSTCore/interface/alpaka/HitsDeviceCollection.h","RecoTracker/LSTCore/interface/alpaka/HitsRangesDeviceCollection.h","RecoTracker/LSTCore/interface/alpaka/LST.h","RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h","RecoTracker/LSTCore/src/alpaka/Hit.h","RecoTracker/LSTCore/src/alpaka/Kernels.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.h","RecoTracker/LSTCore/src/alpaka/MiniDoublet.h","RecoTracker/LSTCore/src/alpaka/PixelQuintuplet.h","RecoTracker/LSTCore/src/alpaka/PixelTriplet.h","RecoTracker/LSTCore/src/alpaka/Segment.h","RecoTracker/LSTCore/src/alpaka/TrackCandidate.h","RecoTracker/LSTCore/standalone/bin/lst.cc","RecoTracker/LSTCore/standalone/code/core/AccessHelper.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.h","RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc"],"squashed":false,"time":1741981285},"b680eac8c778bb3a386ed845bc92977225b84275":{"files":["DQM/TrackingMonitorSource/python/TrackingSourceConfig_Tier0_cff.py","HLTrigger/Configuration/python/HLT_75e33/modules/hltPixelSeedInputLST_cfi.py","RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LST/plugins/LSTOutputConverter.cc"],"squashed":false,"time":1743687617},"c202fe6e6ae21664bd3fe79ab706db6e2c3e8998":{"files":["RecoTracker/LST/interface/LSTPhase2OTHitsInput.h","RecoTracker/LST/interface/LSTPixelSeedInput.h","RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LST/plugins/LSTPhase2OTHitsInputProducer.cc","RecoTracker/LST/plugins/alpaka/LSTProducer.cc","RecoTracker/LSTCore/interface/HitsHostCollection.h","RecoTracker/LSTCore/interface/HitsRangesHostCollection.h","RecoTracker/LSTCore/interface/HitsSoA.h","RecoTracker/LSTCore/interface/LSTInput.h","RecoTracker/LSTCore/interface/alpaka/HitsDeviceCollection.h","RecoTracker/LSTCore/interface/alpaka/HitsRangesDeviceCollection.h","RecoTracker/LSTCore/interface/alpaka/LST.h","RecoTracker/LSTCore/src/alpaka/Hit.h","RecoTracker/LSTCore/src/alpaka/LST.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.dev.cc","RecoTracker/LSTCore/src/alpaka/LSTEvent.h","RecoTracker/LSTCore/src/alpaka/Segment.h","RecoTracker/LSTCore/standalone/bin/lst.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.cc","RecoTracker/LSTCore/standalone/code/core/trkCore.h","RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc"],"squashed":false,"time":1741711024},"c7f7e95dae19b9d53227028e2f5129eaa127f302":{"files":["RecoTracker/LST/plugins/LSTInputProducer.cc","RecoTracker/LST/plugins/LSTOutputConverter.cc","RecoTracker/LSTCore/BuildFile.xml","RecoTracker/LSTCore/interface/LSTInputSoA.h","RecoTracker/LSTCore/interface/LSTPrepareInput.h"],"squashed":false,"time":1742998500},"e9ef704874849e68bd5a8eacd21d7a4b89ee8462":{"files":["RecoTracker/LST/src/classes.h","RecoTracker/LST/src/classes_def.xml"],"squashed":false,"time":1742387180}},"emoji":{},"last_seen_sha":"799c1937743184ee6a0e8d2a5b37997eafd90a61","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44377

- Found files with invalid states:
  - RecoTracker/LSTCore/interface/LSTInput.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Modified: b39f14dccdb2e79ee802a2c46e0325db31ef9150, 38eca3068bf7842ca155d5e6894a191af0074aef, 0bb4f24862a5cd6f76b923387166e5931d6e5bdb
    - Deleted: 5cc987239cae28df23baaeb1855de57a2ac5fb27
  - RecoTracker/LSTCore/interface/alpaka/HitsRangesDeviceCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/HitsRangesHostCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h:
    - Added: b39f14dccdb2e79ee802a2c46e0325db31ef9150
    - Deleted: 0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44377/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44377/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44377/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44377/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44378

- Found files with invalid states:
  - RecoTracker/LSTCore/interface/LSTInput.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Modified: b39f14dccdb2e79ee802a2c46e0325db31ef9150, 38eca3068bf7842ca155d5e6894a191af0074aef, 0bb4f24862a5cd6f76b923387166e5931d6e5bdb
    - Deleted: 5cc987239cae28df23baaeb1855de57a2ac5fb27
  - RecoTracker/LSTCore/interface/alpaka/HitsRangesDeviceCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/HitsRangesHostCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h:
    - Added: b39f14dccdb2e79ee802a2c46e0325db31ef9150
    - Deleted: 0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824

cmsbuild commented:
A new Pull Request was created by  @ariostas for master.

It involves the following packages:

- DQM/TrackingMonitorSource (**dqm**)
- HLTrigger/Configuration (**hlt**)
- RecoTracker/IterativeTracking (**reconstruction**)
- RecoTracker/LST (**reconstruction**)
- RecoTracker/LSTCore (**reconstruction**)


@Martin-Grunewald, @antoniovagnerini, @cmsbuild, @jfernan2, @mandrenguyen, @mmusich, @rseidita can you please review it and eventually sign? Thanks.
@GiacomoSguazzoni, @Martin-Grunewald, @SohamBhattacharya, @VinInn, @VourMa, @arossi83, @dgulhan, @felicepantaleo, @fioriNTU, @gpetruc, @idebruyn, @jandrea, @missirol, @mmusich, @mtosi, @richa2710, @rovere, @sroychow, @threus this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

ariostas commented:
Hmm I see that now there's #47697 and #47774, so these are all likely related issues

slava77 commented:
test parameters:
- enable_tests = gpu
- workflows_gpu = 29634.704,29834.704
- workflows = 29634.703,29834.703
- relvals_opt = -w upgrade,standard
- relvals_opt_gpu = -w upgrade,standard

slava77 commented:
@cmsbuild please test

VourMa commented:
@slava77 I would propose we start testing also the .755 workflow to make sure that LST in Phase 2 HLT works fine.
In this particular case, I expect that workflow to crash in more than one ways, which I can follow up on and resolve once the rest of the issues are fixed.

slava77 commented:
> @slava77 I would propose we start testing also the .755 workflow to make sure that LST in Phase 2 HLT works fine. In this particular case, I expect that workflow to crash in more than one ways, which I can follow up on and resolve once the rest of the issues are fixed.

I don't see an accelerator specifier in this workflow.
Is it expected to be added to the CPU or GPU list?

VourMa commented:
> Is it expected to be added to the CPU or GPU list?

It should be in the "workflows" (not the "workflows_gpu") list.

cmsbuild commented:
-1

**Failed Tests**: RelVals-CUDA RelVals-ROCM 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45418/summary.html
**COMMIT**: 650b6436fbaabe0c2f165014bb671710ce76fafa
**CMSSW**: CMSSW_15_1_X_2025-04-07-1100/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47793/45418/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-07-1100+a5bdf1/67868/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 323 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-07-1100+a5bdf1/67868/validateJR.html)
 - DQMHistoTests: Total files compared: 54
 - DQMHistoTests: Total histograms compared: 4206802
 - DQMHistoTests: [Total failures: 7529](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-07-1100+a5bdf1/67868/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4199253
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 53 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 231 log files, 200 edm output root files, 54 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-07-1100+a5bdf1/67868/triggerResults): no differences found

makortel commented on RecoTracker/LSTCore/src/classes_def.xml:
```diff
@@ -0,0 +1,18 @@
+<lcgdict>
+  <class name="lst::Params_pLS::ArrayUxHits"/>
```
Since this type is really `edm::StdArray<unsigned int, 4>`, it should be placed in `DataFormats/Common/src/classes_def.xml` (either in this or in a separate PR).

slava77 commented:
> **Failed Tests**: RelVals-CUDA RelVals-ROCM

in CUDA [.704 workflow](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45418/runTheMatrixCUDA-results/29634.704_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly/step3_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly.log): 

```
RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h:6:
.../DataFormats/Portable/interface/alpaka/PortableCollection.h:17:54: error: use of undeclared identifier 'Device'
  using PortableCollection = ::PortableCollection<T, Device>;
                                                     ^
...
Error parsing payload code for class PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > with content:
```

IIUC, this is "expected" from local tests, but feedback is needed on how to resolve this

makortel commented on RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h:
```diff
@@ -0,0 +1,19 @@
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+  using LSTInputCollection = std::conditional_t<std::is_same_v<Device, alpaka::DevCpu>,
+                                                ::lst::LSTInputHostCollection,
+                                                ::lst::LSTInputDeviceCollection<Device>>;
```
I wonder if
```suggestion
  using LSTInputCollection = std::conditional_t<std::is_same_v<ALPAKA_ACCELERATOR_NAMESPACE::Device, alpaka::DevCpu>,
                                                ::lst::LSTInputHostCollection,
                                                ::lst::LSTInputDeviceCollection<ALPAKA_ACCELERATOR_NAMESPACE::Device>>;
```
would make any difference wrt. the auto-parsing failure

ariostas replied:
I tried this, and it gets replaced by this error:
```
error: no member named 'Device' in namespace 'ALPAKA_ACCELERATOR_NAMESPACE'
```
So the issue is that it's not including `HeterogeneousCore/AlpakaInterface/interface/config.h` correctly.

makortel replied:
Ok, thanks for the test. The leading order problem is nevertheless the header parsing itself, we'd want ROOT to not do that. I'm afraid we need to investigate more deeply what ROOT ends up doing to cause the header parsing.

@pcanal

slava77 replied:
Is there something "original/creative" in this file? I thought that non-persisted heterogeneous data handling via `edm::Event` is supposed to work.

makortel replied:
No, by eye this file and the `classes_def.xml` look fine (or, similar-enough to other cases that have not caused problems).

Ok, looking at the aforementioned [log](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45418/runTheMatrixCUDA-results/29634.704_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly/step3_TTbar_14TeV+Run4D110_lstOnGPUIters01TrackingOnly.log) again I noticed the error message was about
```
edm::Wrapper<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
``` 

With `PortableDevice...` there should be one layer of `edm::DeviceProduct` between it and the `edm::Wrapper` (as is in the `classes_def.xml`), i.e.
```
edm::Wrapper<edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > > >
```

Then I'd guess this line
https://github.com/cms-sw/cmssw/blob/650b6436fbaabe0c2f165014bb671710ce76fafa/RecoTracker/LST/plugins/alpaka/LSTProducer.cc#L69
should be
```
    device::EDGetToken<lst::LSTInputCollection> lstInputToken_;
```
Ideally this should have resulted a `ProductNotFound` exception instead.

ariostas replied:
Thank you Matti! That fixed the header parsing issue. However, now I'm back to a more standard error:
```
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
Looking for module label: lstInputProducer
Looking for productInstanceName: 
```
One thing that is not clear to me is whether I need to instantiate the `CopyToDevice<LSTInputCollection>` struct somewhere or do something else so that the host product can be fetched as a device product. Is there some way to print the available products to see where the mismatch might be?

makortel replied:
> ```
> Exception Message:
> Principal::getByToken: Found zero products matching all criteria
> Looking for type: edm::DeviceProduct<PortableDeviceMultiCollection<alpaka::DevUniformCudaHipRt<alpaka::ApiCudaRt>,lst::InputHitsSoALayout<128,false>,lst::InputPixelHitsSoALayout<128,false>,lst::InputPixelSeedsSoALayout<128,false> > >
> Looking for module label: lstInputProducer
> Looking for productInstanceName: 
> ```

This is produced by the `LSTInputProducer`, right? In that case, in order to benefit from the implicit host-to-device copy, you have to make the `LSTInputProducer` to an Alpaka module. This behavior is kind-of implied in https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/AlpakaCore/README.md#edproducer, although the point of non-Alpaka modules not taking part in the implicit copies is not explicitly mentioned.

ariostas replied:
Ah okay, I didn't realize it also had to be an Alpaka module, but that makes sense. Thanks!

makortel replied:
An alternative to the `conditional_t` would be
```suggestion
  using LSTInputCollection = PortableMultiCollection<InputHitsSoA, InputPixelHitsSoA, InputPixelSeedsSoA>;
```
Then `RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h` could be removed.

Actually some other places follow a pattern where the aforementioned type alias would be called `LSTInputDeviceCollection`.

(I'm not sure if it would need `PortableMultiCollection3` for the "dictionary compilation to succeed")

ariostas replied:
Thank you Matti. This issue is fixed now. There's a segfault happening at a later point, but I'm looking into it.

ariostas replied:
> There's a segfault happening at a later point, but I'm looking into it

Nevermind, I was just doing something wrong. I think this should be fine now

slava77 commented:
> Judging from the [logs](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45418/runTheMatrix-results/) wf 29834.755 does not seem to have run (probably it was added to the list of commands too late).

yes, it was added after the initial test.
Considering this PR was expected to have a problem with the dictionaries, I didn't run new tests.

makortel commented on RecoTracker/LSTCore/interface/LSTInputHostCollection.h:
```diff
@@ -0,0 +1,11 @@
+#include "DataFormats/Portable/interface/PortableHostCollection.h"
+
+namespace lst {
+  // This needs to be PortableHostCollection3 instead of PortableHostMultiCollection for it to work
```
Does "to work" mean that otherwise the ROOT dictionary compilation fails?

ariostas replied:
Yeah, it doesn't make sense since they should be identical.
https://github.com/cms-sw/cmssw/blob/e3e760917f4c3383bf09f78b6fd220bd2911a27f/DataFormats/Portable/interface/PortableHostCollection.h#L344
But if I define it as
```
using LSTInputHostCollection = PortableHostMultiCollection<InputHitsSoA, InputPixelHitsSoA, InputPixelSeedsSoA>;
```
Then I get this error
```
<artificial>:(.text.startup+0x6dd): undefined reference to `ROOT::GenerateInitInstance(PortableHostMultiCollection<lst::InputHitsSoALayout<128ul, false>, lst::InputPixelHitsSoALayout<128ul, false>, lst::InputPixelSeedsSoALayout<128ul, false> > const*)'
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44465

- Found files with invalid states:
  - RecoTracker/LSTCore/interface/LSTInput.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Modified: b39f14dccdb2e79ee802a2c46e0325db31ef9150, 38eca3068bf7842ca155d5e6894a191af0074aef, 0bb4f24862a5cd6f76b923387166e5931d6e5bdb
    - Deleted: 5cc987239cae28df23baaeb1855de57a2ac5fb27
  - RecoTracker/LSTCore/interface/LSTInputDeviceCollection.h:
    - Added: 0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824
    - Modified: 799c1937743184ee6a0e8d2a5b37997eafd90a61
    - Deleted: 0d72544a608c959441039752c86f6b88682af626
  - RecoTracker/LST/plugins/LSTInputProducer.cc:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Modified: b39f14dccdb2e79ee802a2c46e0325db31ef9150, 38eca3068bf7842ca155d5e6894a191af0074aef, 0bb4f24862a5cd6f76b923387166e5931d6e5bdb, 5cc987239cae28df23baaeb1855de57a2ac5fb27, 69a896d6bafbe4aeba5934126a6b3788be1a09cf, c7f7e95dae19b9d53227028e2f5129eaa127f302, b680eac8c778bb3a386ed845bc92977225b84275
    - Deleted: 0d72544a608c959441039752c86f6b88682af626
  - RecoTracker/LSTCore/interface/alpaka/LSTInputCollection.h:
    - Added: 0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824
    - Modified: 799c1937743184ee6a0e8d2a5b37997eafd90a61
    - Deleted: 0d72544a608c959441039752c86f6b88682af626
  - RecoTracker/LSTCore/interface/alpaka/HitsRangesDeviceCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/HitsRangesHostCollection.h:
    - Added: c202fe6e6ae21664bd3fe79ab706db6e2c3e8998
    - Deleted: b39f14dccdb2e79ee802a2c46e0325db31ef9150
  - RecoTracker/LSTCore/interface/alpaka/LSTInputDeviceCollection.h:
    - Added: b39f14dccdb2e79ee802a2c46e0325db31ef9150, 0d72544a608c959441039752c86f6b88682af626
    - Deleted: 0ba4c297d0d87f9e51f5ee6ac7e2e942a343e824

jfernan2 commented:
assign heterogeneous

cmsbuild commented:
New categories assigned: heterogeneous

@fwyzard,@makortel you have been requested to review this Pull request/Issue and eventually sign? Thanks

slava77 commented:
@cmsbuild please test

VourMa committed:
Make LSTInputProducer work in HLT

ariostas committed:
Refactored LST inputs

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44518

cmsbuild commented:
Pull request #47793 was updated. @Martin-Grunewald, @antoniovagnerini, @cmsbuild, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @mmusich, @rseidita can you please check and sign again.

ariostas committed:
Fixed issues with LST inputs

Moved trackFromSeedFitFailed.h to DataFormats/TrackReco

Addressed review comments: cleanup and pinned memory for LSTInputHostCollection

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44535

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629

cmsbuild commented:
Pull request #47793 was updated.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-CUDA 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45640/summary.html
**COMMIT**: 848d7da2a387c288a81c4fe742b26d42fcf4d603
**CMSSW**: CMSSW_15_1_X_2025-04-18-1100/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47793/45640/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-CUDA

- [12834.406](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45640/runTheMatrixCUDA-results/12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka)```12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka.log```
- [12834.403](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45640/runTheMatrixCUDA-results/12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation)```12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation.log```
- [12834.402](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45640/runTheMatrixCUDA-results/12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka)```12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka.log```


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68057/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 233 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68057/validateJR.html)
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4216615
 - DQMHistoTests: [Total failures: 4583](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68057/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4212012
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 54 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 234 log files, 202 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68057/triggerResults): found differences in 1 / 53 workflows


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68058/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-18-1100+a5bdf1/68058/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

ariostas commented:
It looks to me like the errors are unrelated, since the LST-specific workflows ran successfully. They're likely related to #47808.

I'll add an extra commit that stemmed from internal discussions, but apart from that this PR is ready to be reviewed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44562

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629
  - File  HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepTrackCandidates_cfi.py  modified in PR(s): #47891
  - File  HLTrigger/Configuration/python/HLT_75e33/sequences/HLTInitialStepSequence_cfi.py  modified in PR(s): #47891

slava77 commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-CUDA 
**Size**: This PR adds an extra 96KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45643/summary.html
**COMMIT**: e8bd3432ef63f6c3223914f6fbc994bab77fd69b
**CMSSW**: CMSSW_15_1_X_2025-04-18-2300/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47793/45643/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-CUDA

- [12834.406](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45643/runTheMatrixCUDA-results/12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka)```12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka.log```
- [12834.403](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45643/runTheMatrixCUDA-results/12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation)```12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation.log```
- [12834.402](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45643/runTheMatrixCUDA-results/12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka)```12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka.log```


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68061/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68061/validateJR.html)
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4216615
 - DQMHistoTests: [Total failures: 38](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68061/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4216557
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 54 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 234 log files, 202 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68061/triggerResults): found differences in 1 / 53 workflows


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68062/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-18-2300+a5bdf1/68062/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

slava77 commented:
@fwyzard @makortel @cms-sw/heterogeneous-l2 
please clarify on the status of the GPU workflows: crashes in RelVals-CUDA are not in workflows directly related to this PR.
Should we wait until these are fixed or can we proceed with the review?
Thank you.

fwyzard commented:
Personally I just got back from 2 weeks away (work + vacation), so I have accumulated quite a large back log.
I will eventually get to reviewing this PR, probably in a couple weeks' time.

slava77 commented:
> Personally I just got back from 2 weeks away (work + vacation), so I have accumulated quite a large back log. I will eventually get to reviewing this PR, probably in a couple weeks' time.

Thank you for the update and clarification of the timeline.

@makortel 
Would you be able to review and sign sooner?
Please let me know.
Thank you.

fwyzard commented:
ignore tests-rejected with ib-failure

makortel commented on RecoTracker/LST/plugins/alpaka/LSTInputProducer.cc:
```diff
@@ -0,0 +1,225 @@
+        beamSpotToken_(consumes(iConfig.getParameter<edm::InputTag>("beamSpot"))),
+        lstPixelSeedsPutToken_(produces()),
+        lstInputPutToken_(produces()) {
+    seedTokens_ =
+        edm::vector_transform(iConfig.getParameter<std::vector<edm::InputTag>>("seedTracks"),
+                              [&](const edm::InputTag& tag) { return consumes<edm::View<reco::Track>>(tag); });
```
This could be done as part of the initializer list, in which case `seedTokens_` could be made `const`.

makortel commented on RecoTracker/LST/plugins/alpaka/LSTInputProducer.cc:
```diff
@@ -0,0 +1,225 @@
+    std::vector<std::vector<int>> see_hitIdx;
+    TrajectorySeedCollection see_seeds;
+
+    for (size_t iColl = 0; iColl < seedTokens_.size(); ++iColl) {
+      // Get seed tokens
+      auto const& seedToken = seedTokens_[iColl];
```
Looks like this could be
```suggestion
    for (auto const& seedToken : seedTokens_) {
```

makortel commented on RecoTracker/LST/plugins/alpaka/LSTInputProducer.cc:
```diff
@@ -0,0 +1,225 @@
+        see_stateTrajGlbPy.push_back(stateGlobal.momentum().y());
+        see_stateTrajGlbPz.push_back(stateGlobal.momentum().z());
+        see_q.push_back(seedTrack.charge());
+        see_hitIdx.push_back(hitIdx);
```
```suggestion
        see_hitIdx.push_back(std::move(hitIdx));
```
or
```suggestion
        see_hitIdx.emplace_back(std::move(hitIdx));
```
might avoid a copy of `hitIdx`.

makortel commented on Validation/RecoTrack/plugins/TrackFromSeedProducer.cc:
Since this plugin is now used outside of the `Validation` subsystem, it should be moved to (e.g.) some reconstruction package. But perhaps that's better to be done in a separate PR.

makortel reviewed (commented):
On a cursory look this PR looks ok to me. I left couple of suggestions that could be polished here or in a subsequent PR.

makortel commented:
+core

The `DataFormats/Common` part is good in any case.

jfernan2 commented:
+1

ariostas commented:
Thank you, @makortel and @fwyzard!

@makortel I'll apply the changes you suggested.

@fwyzard:

1. Yeah, that's a good point. I initially was writing it as a non-alpaka plugin, but now that it was made an alpaka plugin, it makes sense to take more advantage of it.
2. There is no direct vector to vector copying as there is some processing going on. But this could definitely be disentangled and made better in a future PR.
3. I'm planning to work on the outputs in another PR. This PR is already pretty sizable as it is, and there are also other output-related things that we promised to address in #46746.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44620

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629
  - File  HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepTrackCandidates_cfi.py  modified in PR(s): #47891
  - File  HLTrigger/Configuration/python/HLT_75e33/sequences/HLTInitialStepSequence_cfi.py  modified in PR(s): #47891

cmsbuild commented:
Pull request #47793 was updated. @Martin-Grunewald, @antoniovagnerini, @cmsbuild, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @mmusich, @rseidita can you please check and sign again.

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-CUDA 
**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45722/summary.html
**COMMIT**: 82897120464dc6489d70b97376c6ad19335ee00b
**CMSSW**: CMSSW_15_1_X_2025-04-25-1100/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47793/45722/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-CUDA

- [12834.406](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45722/runTheMatrixCUDA-results/12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka)```12834.406_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyTripletsAlpaka.log```
- [12834.403](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45722/runTheMatrixCUDA-results/12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation)```12834.403_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka_Validation.log```
- [12834.402](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45722/runTheMatrixCUDA-results/12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka)```12834.402_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka/step3_TTbar_14TeV+2024_Patatrack_PixelOnlyAlpaka.log```


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68145/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68145/validateJR.html)
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4213551
 - DQMHistoTests: [Total failures: 142](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68145/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4213389
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 54 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 234 log files, 202 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68145/triggerResults): found differences in 1 / 53 workflows


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68146/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-04-25-1100+a5bdf1/68146/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

fwyzard commented:
ignore tests-rejected with ib-failure

ariostas commented:
@fwyzard @makortel @jfernan2 If you don't have further comments, could you sign on this when you have a chance? Thanks!

jfernan2 commented:
+1

VourMa commented:
Unfortunately, with the merging of #47891, there are a couple changes that need to be done for this PR to fully work, more specifically:
https://github.com/cms-sw/cmssw/blob/a31424ec6d9989e1390981466d09b137ad068318/HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepTrajectorySeedsLST_cfi.py#L5-L6
and
https://github.com/cms-sw/cmssw/blob/a31424ec6d9989e1390981466d09b137ad068318/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTInitialStepSequence_cfi.py#L58-L60
need to be updated.

These are newly added files, hence it seems as if there is no conflict. I can fix this a bit later today, and push the commits.

VourMa committed:
Adjust HLT config with LST seeding

Cleanup in the HLT part

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44665

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629
  - File  RecoTracker/LSTCore/src/alpaka/PixelQuintuplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/src/alpaka/PixelTriplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc  modified in PR(s): #47995

cmsbuild commented:
Pull request #47793 was updated. @Martin-Grunewald, @antoniovagnerini, @cmsbuild, @fwyzard, @makortel, @mmusich, @rseidita can you please check and sign again.

VourMa commented:
> Unfortunately, with the merging of #47891, there are a couple changes that need to be done for this PR to fully work, more specifically:
> 
> https://github.com/cms-sw/cmssw/blob/a31424ec6d9989e1390981466d09b137ad068318/HLTrigger/Configuration/python/HLT_75e33/modules/hltInitialStepTrajectorySeedsLST_cfi.py#L5-L6
> 
> and
> 
> https://github.com/cms-sw/cmssw/blob/a31424ec6d9989e1390981466d09b137ad068318/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTInitialStepSequence_cfi.py#L58-L60
> 
> need to be updated.
> 
> These are newly added files/lines, hence it seems as if there is no conflict. I can fix this a bit later today, and push the commits.

The new HLT config should now be functional with this PR. Could we add the `0.757` workflow to cross-check when we rerun the tests? Thanks!

slava77 commented:
test parameters:
- enable_tests = gpu
- workflows_gpu = 29634.704,29834.704
- workflows = 29634.703,29834.703,29834.755,29634.757,29834.757
- relvals_opt = -w upgrade,standard
- relvals_opt_gpu = -w upgrade,standard

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45818/summary.html
**COMMIT**: ed8af8c7f9a14a4f0710e327bdc740561cd065c7
**CMSSW**: CMSSW_15_1_X_2025-05-02-1100/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47793/45818/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68246/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68246/validateJR.html)
 - DQMHistoTests: Total files compared: 57
 - DQMHistoTests: Total histograms compared: 4405457
 - DQMHistoTests: [Total failures: 61](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68246/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4405376
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 56 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 240 log files, 206 edm output root files, 57 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68246/triggerResults): found differences in 3 / 55 workflows


## CUDA Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68240/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68240/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68245/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-02-1100+a5bdf1/68245/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

ariostas commented on HLTrigger/Configuration/python/HLT_75e33/sequences/HLTInitialStepSequence_cfi.py:
```diff
@@ -27,9 +26,8 @@
-    +hltPixelSeedInputLST
     +hltSiPhase2RecHits # Probably need to move elsewhere in the final setup
-    +hltPhase2OTHitsInputLST # Probably need to move elsewhere in the final setup
+    +hltInputLST # Probably need to move elsewhere in the final setup
```
@VourMa?

fwyzard commented:
> * Yeah, that's a good point. I initially was writing it as a non-alpaka plugin, but now that it was made an alpaka plugin, it makes sense to take more advantage of it.

Thanks for the update.

> * There is no direct vector to vector copying as there is some processing going on. But this could definitely be disentangled and made better in a future PR.

OK.

> * I'm planning to work on the outputs in another PR. This PR is already pretty sizable as it is, and there are also other output-related things that we promised to address in

OK.

fwyzard commented:
+heterogeneous

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44726

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629
  - File  RecoTracker/LSTCore/src/alpaka/PixelQuintuplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/src/alpaka/PixelTriplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc  modified in PR(s): #47995

mmusich commented:
+hlt

cmsbuild commented:
Pull request #47793 was updated. @antoniovagnerini, @cmsbuild, @fwyzard, @makortel, @rseidita can you please check and sign again.

VourMa commented:
Thanks for the review, @mmusich! All of your comments have been addressed in the last commit.

@fwyzard Sorry but it seems like the bot also removed your signature with the last push (I thought it wouldn't do it, since only HLT was touched).

mmusich commented:
> All of your comments have been addressed in the last commit.

thanks and sorry for the belated review.

> Sorry but it seems like the bot also removed your signature with the last push (I thought it wouldn't do it, since only HLT was touched).

@smuzaffar can you have a look?

mmusich commented:
@cmsbuild, please test

makortel commented:
> > Sorry but it seems like the bot also removed your signature with the last push (I thought it wouldn't do it, since only HLT was touched).
> 
> @smuzaffar can you have a look?

IIRC this happens because the `heterogeneous` category was explicitly assigned rather than assigned automatically via the touched packages. Signatures for explicitly assigned categories are reset on every update.

fwyzard commented:
> Sorry but it seems like the bot also removed your signature with the last push (I thought it wouldn't do it, since only HLT was touched).

No problem.

fwyzard commented:
+heterogeneous

fwyzard commented:
Actually, could you squash the last commits into the previous one ?

cmsbuild commented:
+1

**Size**: This PR adds an extra 44KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45889/summary.html
**COMMIT**: b5d3a4c42bd98a308ddf6b87a382b911bf693c34
**CMSSW**: CMSSW_15_1_X_2025-05-06-1100/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47793/45889/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68310/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 611 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68310/validateJR.html)
 - DQMHistoTests: Total files compared: 57
 - DQMHistoTests: Total histograms compared: 4405457
 - DQMHistoTests: [Total failures: 2443](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68310/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4402994
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 56 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 240 log files, 206 edm output root files, 57 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68310/triggerResults): found differences in 3 / 55 workflows


## CUDA Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68313/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68313/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68309/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-06-1100+a5bdf1/68309/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

VourMa commented:
> Actually, could you squash the last commits into the previous one ?

In the last force push, I simplified the commit history by squashing together a few of the commits.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47793/44732

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/src/classes_def.xml  modified in PR(s): #47629
  - File  RecoTracker/LSTCore/src/alpaka/PixelQuintuplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/src/alpaka/PixelTriplet.h  modified in PR(s): #47995
  - File  RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc  modified in PR(s): #47995

cmsbuild commented:
Pull request #47793 was updated. @antoniovagnerini, @cmsbuild, @rseidita can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

cmsbuild commented:
+1

**Size**: This PR adds an extra 44KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a5bdf1/45907/summary.html
**COMMIT**: ecd1aed8b2539511364e40c8f77c799b7c58148f
**CMSSW**: CMSSW_15_1_X_2025-05-06-2300/el8_amd64_gcc12
**Additional Tests**: CUDA,ROCM
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47793/45907/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 198 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68324/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68324/validateJR.html)
 - DQMHistoTests: Total files compared: 57
 - DQMHistoTests: Total histograms compared: 4405457
 - DQMHistoTests: [Total failures: 41](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68324/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4405396
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.316 KiB( 56 files compared)
 - DQMHistoSizes: changed ( 29634.703,... ): -0.158 KiB       Tracking/TrackParameters
 - Checked 240 log files, 206 edm output root files, 57 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68324/triggerResults): found differences in 3 / 55 workflows


## CUDA Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68332/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsCUDA/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68332/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files


## ROCM Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68325/validateJR.html)
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsROCM/CMSSW_15_1_X_2025-05-06-2300+a5bdf1/68325/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @rappoccio, @mandrenguyen, @antoniovilela (and backports should be raised in the release meeting by the corresponding L2)
